designs/filewriter-refactor.md

```diff
@@ -0,0 +1,100 @@
+
+With respect to the `FileWriter` interface composition, `FileWriter` defines all methods for the concerns outlined earlier directly. This forces consumers to depend on the full `FileWriter` interface even if they do not need to define new path contexts. The vast majority of consumers only care about writing files hence the other 4 methods are superfluous in most contexts.
+
+Finally, the `Write()` method writes files with unix permission bits 0777 before the umask is applied. This brings about risk as files are written with the execution bit for both group and other assuming a default umask of 022.
+
+## Proposal
+
+I propose we reduce the scope of the `FileWriter` to simply providing convinience methods for writing under a path context. Reducing the scope includes removal of the `CleanUp()` and `CleanUpTemp()` methods and any persistent or temporary file options. Additionally, we make clear consumers are responsible for cleaning any files they create with code docs on creation methods. If consumers wish to delete the whole directory they can use the standard libraries `os.RemoveAll(w.Dir())`. Additionally, any files created are assumed to be permanent. Consumers can explicitly define temporary file path contexts that are managed as independent contexts if they need to model temporary files. Finally, I propose a breakdown of interfaces to allow consumers to depend on only what they need and an introduction of a method set that compliments the standard libraries `io` package. Complienting the standard libraries `io` package ensures consumers can better satisfy the SRP as they can function without needing to manage file properties such as name and permissions.
+
+With respect to the default permissions assigned to files, I propose we change to using 0644 granting read and write to the owner and read permissions to group and other.
+
+## Implementation
+
+The following interfaces accept names as at least 1 parameter. The default implementation will prepend the path context to the name and return the `*os.File` type that can be used in `io` interfaces such as `io.Reader` and `io.Writer`.
+
+```go
+// Opener has the same behavior as os.Open().
+type Opener interface {
```

Comment:
I'm not 100% because I haven't evaluated the code that reads files but given we do a lot of file reading considering the amount of writing, I suspect it would be useful.